### PR TITLE
Sampling (and sub-sampling emulation) / Removal of done/stop channels.

### PR DESCRIPTION
**Pull request by ycamper**

### CODE REVIEW

File: cmd/client.go [hunk 1]
```diff
@@ -18,7 +17,9 @@ import (
 	"github.com/criticalstack/swoll/pkg/event/call"
 	color "github.com/fatih/color"
 	uuid "github.com/google/uuid"
+	log "github.com/sirupsen/logrus"
 	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v2"
```
Comment (ktravis) [inline cmd/client.go:22]: May want to use `sigs.k8s.io/yaml` instead here, for consistency.

File: internal/pkg/hub/hub.go [hunk 2]
```diff
@@ -253,6 +257,28 @@ func (h *Hub) PushJob(job *Job, ns, nr int) {
...
+	if joblist == nil {
+		return nil
+	}
+
+	min := joblist.Front().Value.(*JobContext)
+
+	var next *list.Element
+
+	for j := joblist.Front(); j != nil; j = next {
+		next = j.Next()
```
Comment (ktravis) [inline internal/pkg/hub/hub.go:271]: Any reason this isn't:
```go
for j := joblist.Front(); j != nil; j = j.Next() {
```
?

File: internal/pkg/hub/job.go [hunk 3]
```diff
@@ -140,48 +149,115 @@ func (j *Job) Run(h *Hub, done chan bool) error {
...
 			return nil
 		}
 	}
 }
 
 // WriteEvent writes event `ev` to all listeners of this `Job`
 func (j *Job) WriteEvent(h *Hub, ev *event.TraceEvent) {
-	h.ps.Publish(ev, pubsub.LinearTreeTraverser(
-		hashPath(swJobStream, j.JobID())))
+	if j.Spec.SampleRate > 0 {
```
Comment (ktravis) [inline internal/pkg/hub/job.go:214]: This function might be a little clearer as an early return in the case of no sampling - but again, personal preference.

**Review by ktravis — CHANGES_REQUESTED** [review]
Looks great to me. Most things were just notes or small questions, only thing I would suggest is determining whether to use sigs.k8s.io/yaml or not. If the answer is to stick with gopkg.in/yaml.v2, just let me know and I'll switch this to approved.

File: cmd/client.go [hunk 4]
```diff
@@ -86,44 +88,44 @@ var cmdClientWatch = &cobra.Command{
...
 
-func watchJob(endpoints []*client.Endpoint, id string, stopChan chan bool) {
-	ctx := context.Background()
-
+func watchJob(ctx context.Context, endpoints []*client.Endpoint, id string) {
 	outChan := make(chan *client.StreamMessage)
 	defer close(outChan)
 	wg := sync.WaitGroup{}
 
+	ctx, cancel := context.WithCancel(ctx)
```
[the remark this answers is not in the document]
Comment (ycamper) [reply cmd/client.go:116]: I figured it was good for propagation later on as I can use the context to timeout websocket reads and writes at a later point.

File: internal/pkg/hub/job.go (hunk 3 above)
Comment (ycamper) [reply internal/pkg/hub/job.go:214]: I honestly don't care which conditional comes first - I've always done more-complex > least-complex.

File: cmd/client.go [hunk 5]
```diff
@@ -243,10 +245,10 @@ var cmdClientCreate = &cobra.Command{
...
 
-		ctx := context.Background()
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
 
 		outChan := make(chan *client.StreamMessage)
-		stpChan := make(chan bool)
 		sigChan := make(chan os.Signal, 1)
 
 		signal.Notify(sigChan, os.Interrupt, syscall.SIGTERM)
```
[the remark this answers is not in the document]
Comment (ycamper) [reply cmd/client.go:253]: fancy. updated.

**Review by ktravis — APPROVED** [review]

**Pull request merged**